hwilib/devices/ledger.py

```diff
@@ -117,209 +147,272 @@ def func(*args: Any, **kwargs: Any) -> Any:
-                has_legacy = True
-
-            if psbt_in.non_witness_utxo and use_trusted_segwit:
-                ledger_prevtx = bitcoinTransaction(psbt_in.non_witness_utxo.serialize())
-                segwit_inputs[-1].update(self.app.getTrustedInput(ledger_prevtx, txin.prevout.n))
-
-            pubkeys = []
-            signature_attempts = []
-
-            # Save scriptcode for later signing
-            script_codes[i_num] = scriptcode
-
-            # Find which pubkeys could sign this input (should be all?)
-            for pubkey in psbt_in.hd_keypaths.keys():
-                if hash160(pubkey) in scriptcode or pubkey in scriptcode:
-                    pubkeys.append(pubkey)
-
-            # Figure out which keys in inputs are from our wallet
-            for pubkey in pubkeys:
-                keypath = psbt_in.hd_keypaths[pubkey]
-                if master_fpr == keypath.fingerprint:
-                    # Add the keypath strings
-                    keypath_str = keypath.get_derivation_path()[2:] # Drop the leading m/
-                    signature_attempts.append((keypath_str, pubkey))
-
-            all_signature_attempts[i_num] = signature_attempts
-
-        # Sign any segwit inputs
-        if has_segwit:
-            # Process them up front with all scriptcodes blank
-            blank_script_code = bytearray()
-            for i in range(len(segwit_inputs)):
-                self.app.startUntrustedTransaction(i == 0, i, segwit_inputs, script_codes[i] if use_trusted_segwit else blank_script_code, c_tx.nVersion)
-
-            # Number of unused fields for Nano S, only changepath and transaction in bytes req
-            self.app.finalizeInput(b"DUMMY", -1, -1, change_path, tx_bytes)
-
-            # For each input we control do segwit signature
-            for i in range(len(segwit_inputs)):
-                for signature_attempt in all_signature_attempts[i]:
-                    self.app.startUntrustedTransaction(False, 0, [segwit_inputs[i]], script_codes[i], c_tx.nVersion)
-                    tx.inputs[i].partial_sigs[signature_attempt[1]] = self.app.untrustedHashSign(signature_attempt[0], "", c_tx.nLockTime, 0x01)
-        elif has_legacy:
-            first_input = True
-            # Legacy signing if all inputs are legacy
-            for i in range(len(legacy_inputs)):
-                for signature_attempt in all_signature_attempts[i]:
-                    assert(tx.inputs[i].non_witness_utxo is not None)
-                    self.app.startUntrustedTransaction(first_input, i, legacy_inputs, script_codes[i], c_tx.nVersion)
-                    self.app.finalizeInput(b"DUMMY", -1, -1, change_path, tx_bytes)
-                    tx.inputs[i].partial_sigs[signature_attempt[1]] = self.app.untrustedHashSign(signature_attempt[0], "", c_tx.nLockTime, 0x01)
-                    first_input = False
-
-        # Send PSBT back
+                self._switch_to_legacy_client()
+                return legacy_sign_tx()
+
```

Comment:
Wouldn't `_switch_to_legacy_client()` make it impossible to use any non-legacy feature from now on, since it permanently changes `self.client`?
I suppose self.client could be restored after `legacy_sign_tx` returns to avoid that.